### ISSUE TITLE
deal with deleted MPI-1 functions

### ISSUE THREAD

**Issue by jeffhammond** [LLNL/mpiP#6]
From https://sourceforge.net/p/mpip/mailman/message/36511483/:

> I am excited to see mpiP and I hope it will help me debug some MPI performance issues in my application. I am running into issues trying to use it though. I suspect I am doing something basic wrong. Can you please help me?
> 
> If it matters, I have OpenMPI 4 on Ubuntu 18.04, gcc 7.3.
> My OpenMPI is installed system-wide in /usr/local/openmpi. I installed mpiP like this:
> ./configure --prefix=/usr/local/mpiP/ --without-f77 && make && sudo make install
> 
> I have a test file with an MPIReduce call. I’m building it as follows:
> mpicc -g reduce.c -o reduce -L/usr/local/mpiP/lib -lmpiP -lm
> 
> I am unable to successfully build the executable.
> /usr/local/mpiP/lib/libmpiP.a(wrappers.o): In function `mpiPif_MPI_Attr_delete':
> /home/ubuntu/mpiP-3.4.1/wrappers.c:489: undefined reference to `PMPI_Attr_delete'
> /home/ubuntu/mpiP-3.4.1/wrappers.c:489: undefined reference to `PMPI_Attr_delete'
> /usr/local/mpiP/lib/libmpiP.a(wrappers.o): In function `mpiPif_MPI_Attr_get':
> /home/ubuntu/mpiP-3.4.1/wrappers.c:555: undefined reference to `PMPI_Attr_get'
> /home/ubuntu/mpiP-3.4.1/wrappers.c:555: undefined reference to `PMPI_Attr_get'
> /usr/local/mpiP/lib/libmpiP.a(wrappers.o): In function `mpiPif_MPI_Attr_put':
> /home/ubuntu/mpiP-3.4.1/wrappers.c:621: undefined reference to `PMPI_Attr_put'
> /home/ubuntu/mpiP-3.4.1/wrappers.c:621: undefined reference to `PMPI_Attr_put'
> /usr/local/mpiP/lib/libmpiP.a(wrappers.o): In function `mpiPif_MPI_Keyval_create':
> /home/ubuntu/mpiP-3.4.1/wrappers.c:4941: undefined reference to `PMPI_Keyval_create'
> /home/ubuntu/mpiP-3.4.1/wrappers.c:4941: undefined reference to `PMPI_Keyval_create'
> /usr/local/mpiP/lib/libmpiP.a(wrappers.o): In function `mpiPif_MPI_Keyval_free':
> /home/ubuntu/mpiP-3.4.1/wrappers.c:5005: undefined reference to `PMPI_Keyval_free'
> /home/ubuntu/mpiP-3.4.1/wrappers.c:5005: undefined reference to `PMPI_Keyval_free'
> /usr/local/mpiP/lib/libmpiP.a(record_stack.o): In function `mpiPi_RecordTraceBack':
> /home/ubuntu/mpiP-3.4.1/record_stack.c:51: undefined reference to `_Ux86_64_getcontext'
> /home/ubuntu/mpiP-3.4.1/record_stack.c:57: undefined reference to `_ULx86_64_init_local'
> /home/ubuntu/mpiP-3.4.1/record_stack.c:66: undefined reference to `_ULx86_64_step'
> /home/ubuntu/mpiP-3.4.1/record_stack.c:73: undefined reference to `_ULx86_64_step'
> /home/ubuntu/mpiP-3.4.1/record_stack.c:76: undefined reference to `_ULx86_64_get_reg'
> /home/ubuntu/mpiP-3.4.1/record_stack.c:66: undefined reference to `_ULx86_64_step'
> collect2: error: ld returned 1 exit status

This is needed because Open-MPI has implemented the deletion of these symbols.

I am working on a patch now.

**Comment by milthorpe** [comment]
As a (nasty) workaround until this is fixed: one can build a small library of dummy implementations of the MPI-1 functions and link it along with mpiP. e.g.
```
// dummy.cpp
#include <mpi.h>
int PMPI_Attr_put(MPI_Comm comm, int keyval, void *attribute_val) { return 0; }
int PMPI_Attr_get(MPI_Comm comm, int keyval, void *attribute_val, int *flag) { return 0; }
int PMPI_Attr_delete(MPI_Comm comm, int keyval) { return 0; }
int PMPI_Keyval_create(MPI_Copy_function *copy_fn,
   MPI_Delete_function *delete_fn,
  int *keyval, void *extra_state) { return 0; }
int PMPI_Keyval_free(int *keyval) { return 0; }
```

```
$ mpic++ dummy.cpp -fPIC -shared -o libdummy_mpi1.so
$ mpic++ -g application.cpp -L/usr/local/lib/openmpi -lmpiP -lm -lbfd -liberty -lunwind -L. -ldummy_mpi1 -o mpi_launcher
```

**Comment by jeffhammond** [comment]
I think I forgot that I was supposed to finish this patch...

**Comment by cchambreau** [comment]
Addressed with commit [586c790df5dcf9e76d2ac4170374b802f2b4f01f](https://github.com/LLNL/mpiP/commit/586c790df5dcf9e76d2ac4170374b802f2b4f01f)
